README.md
<p align="center">
    <img src="./assets/banner.png" width="500">
</p>

## Droid

Droid is a simple HTTP server that exposes some information about a Tendermint blockchain node.

It allows you to retrieve the node ID, public key, latest block, latest block height, and health status of the node.

## Table of Contents

- [Droid](#droid)
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
  * [Endpoints](#endpoints)
  * [Configuration](#configuration)
- [Contributing](#contributing)

## Requirements

- Go 1.18 or later

## Installation

Clone the repository:

```bash
$ git clone https://github.com/osmosis-labs/droid.git
```

Build the binary:

```bash
make build
```

## Usage

Start the HTTP server:

```bash
make run
```

Open a web browser and go to http://localhost:8080/node_id to retrieve the node ID.

### Endpoints

| Endpoint   | Description                             |
|------------|-----------------------------------------|
| `/node_id` | Retrieve the node ID.                   |
| `/pub_key` | Retrieve the public key of the node.    |
| `/block`   | Retrieve the latest block.              |
| `/height`  | Retrieve the latest block height.       |
| `/health`  | Retrieve the health status of the node. |


### Configuration

By default, the application will look for a `config.yaml` file in the current directory and in the `$HOME/.droid` directory. The file should contain the following structure:

```yaml
rpc_endpoint: http://0.0.0.0:26657
lcd_endpoint: http://0.0.0.0:1317
```

If the `config.yaml` file is not found, the application will create a default one with the above values.

Alternatively, you can specify the endpoint values using environment variables. To do so, set the `rpc_endpoint` and `lcd_endpoint` environment variables with the desired values before running the application. The environment variable values will take precedence over any values in the `config.yaml` file. If the file does not exist, a default one will be created.

## Contributing

Contributions are welcome! Please open a pull request or an issue.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.2] - 2023-07-24

- Fix retry time from 1 minute to 4 minute.
- Add goreleaser
- Update golang to `1.20`

## [0.0.1] - 2023-03-23

- Add node id, pub key, latest block, health check handlers
- Support config file for different node addresses.

[unreleased]: https://github.com/osmosis-labs/droid/compare/v0.0.1...HEAD
[0.0.1]: https://github.com/osmosis-labs/droid/releases/tag/v0.0.1

main.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/sirupsen/logrus"
	"github.com/spf13/viper"
)

var (
	config    Config
	statusURL = config.RPCEndpoint + "/status"
	epochURL  = config.LCDEndpoint + "/osmosis/epochs/v1beta1/epochs"
)

var log = logrus.New()

func GetNodeStatus(statusURL string) NodeStatus {
	var nodeStatus NodeStatus
	var err error

	for i := 0; i < 48; i++ { // Retry for up to 4 minute (48 * 5 seconds)
		resp, err := http.Get(statusURL)
		if err == nil {
			defer resp.Body.Close()
			if err := json.NewDecoder(resp.Body).Decode(&nodeStatus); err == nil {
				return nodeStatus
			}
		}
		if i < 11 {
			log.Info("node connection refused...retrying...")
			// If this is not the last retry, wait for 5 seconds before retrying
			time.Sleep(5 * time.Second)
		}
	}
	log.Info("node refused to connect for 4 minute...exiting program")
	// If all retries fail, return empty node status
	log.Fatalln(err)
	return NodeStatus{}
}

func getLatestBlockFromNodeStatus(status NodeStatus) Block {

	return Block{
		Height: status.Result.SyncInfo.LatestBlockHeight,
		Hash:   status.Result.SyncInfo.LatestBlockHash,
		Time:   status.Result.SyncInfo.LatestBlockTime,
	}
}

// /block handler
func getLatestBlockHandler(w http.ResponseWriter, r *http.Request) {

	status := GetNodeStatus(statusURL)
	latestBlock := getLatestBlockFromNodeStatus(status)

	w.Header().Set("Content-Type", "application/json")
	enc := json.NewEncoder(w)
	err := enc.Encode(latestBlock)

	if err != nil {
		log.Fatalf("Failed to encode block, %s\n", err)
		io.WriteString(w, "Internal Server Error\n")
	}
}

// /height handler
func getLatestBlockHeightHandler(w http.ResponseWriter, r *http.Request) {

	status := GetNodeStatus(statusURL)
	latestBlock := getLatestBlockFromNodeStatus(status)
	_, _ = io.WriteString(w, latestBlock.Height)
}

// /node_id handler
func getNodeIDHandler(w http.ResponseWriter, r *http.Request) {
	status := GetNodeStatus(statusURL)
	_, _ = io.WriteString(w, status.Result.NodeInfo.ID)
}

// /pub_key handler
func getPubKeyHandler(w http.ResponseWriter, r *http.Request) {
	status := GetNodeStatus(statusURL)

	response := map[string]string{
		"@type": "/cosmos.crypto.ed25519.PubKey",
		"key":   status.Result.ValidatorInfo.PubKey.Value,
	}

	data, _ := json.Marshal(response)
	_, _ = io.WriteString(w, string(data))
}

// global cached next epoch start time
var nextEpochStartTime time.Time

// This variable controls how many minutes after epoch we should expect the node to properly respond
// We assume epoch starts 5 minutes before the calculated epoch
var minutesAfterEpoch time.Duration = 35

func calculateEpoch() time.Time {
	failure := time.Time{}
	// Calculate the epoch from the node
	resp, err := http.Get(epochURL)
	if err != nil {
		log.Fatalln(err)
	}

	defer resp.Body.Close()
	bz, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println(err)
		return failure
	}
	if resp.StatusCode != http.StatusOK {
		fmt.Println(err)
		return failure
	}

	// Unmarshall the epoch json into a map from string to an array of interfaces
	var epoch map[string]interface{}
	err = json.Unmarshal(bz, &epoch)
	if err != nil {
		fmt.Println(err)
		return failure
	}

	// extract the "epoch" array from the epoch map
	epochArray := epoch["epochs"].([]interface{})

	// iterate over the array and find the element of the array that has identifier "day"
	for _, epochObject := range epochArray {
		epochObject := epochObject.(map[string]interface{})
		epochIdentifier := epochObject["identifier"].(string)
		//check that the identifier is "day"
		if epochIdentifier != "day" {
			continue
		}
		// extract the "current_epoch_start_time" from the epochObject and parse it as a date
		epochStartTimeString := epochObject["current_epoch_start_time"].(string)
		epochStartTime, err := time.Parse(time.RFC3339, epochStartTimeString)
		if err != nil {
			fmt.Println(err)
			return failure
		}
		// add 24 hours to the epoch start time
		nextEpochStartTime = epochStartTime.Add(24 * time.Hour)
		// subtract 5 minutes
		nextEpochStartTime = nextEpochStartTime.Add(-5 * time.Minute)
		return nextEpochStartTime
	}
	fmt.Println("No epochs found")
	return failure
}

// Function getEpoch2 which queries the epochURL and parses it as json
func isWithinEpoch(currentTime time.Time) bool {
	// Check if we have a cached epoch
	if nextEpochStartTime.IsZero() || currentTime.After(nextEpochStartTime.Add(minutesAfterEpoch*time.Minute)) {
		// if there is no cache or the epoch has ended, recalculate
		nextEpochStartTime = calculateEpoch()
	}
	// otherwise, use the cached value

	if currentTime.After(nextEpochStartTime) && currentTime.Before(nextEpochStartTime.Add(minutesAfterEpoch*time.Minute)) {
		return true
	}
	return false
}

// /health handler
func healthcheckHandler(w http.ResponseWriter, r *http.Request) {
	status := GetNodeStatus(statusURL)
	latestBlock := getLatestBlockFromNodeStatus(status)

	catchingUp := status.Result.SyncInfo.CatchingUp
	secondsSinceLastBlock := int(time.Now().UTC().Sub(latestBlock.Time).Seconds())
	isEpoch := isWithinEpoch(time.Now())

	var healthStatus int
	var healthMessage string

	if catchingUp || (secondsSinceLastBlock >= 60 && !isEpoch) {
		healthStatus = http.StatusServiceUnavailable
		healthMessage = "DOWN"
	} else {
		healthStatus = http.StatusOK
		healthMessage = "UP"
	}

	w.WriteHeader(healthStatus)
	_, _ = io.WriteString(w, fmt.Sprintf("%s\nLatest Block %s (Received %d seconds ago)\n", healthMessage, latestBlock.Height, secondsSinceLastBlock))

	log.WithFields(logrus.Fields{
		"catchingUp":            catchingUp,
		"secondsSinceLastBlock": secondsSinceLastBlock,
		"isEpoch":               isEpoch,
		"healthStatus":          healthStatus,
		"healthMessage":         healthMessage,
	}).Debug("Healthcheck status")
}

func readConfig() error {
	viper.SetConfigName("config")
	viper.SetConfigType("yaml")
	viper.AddConfigPath(".")
	viper.AddConfigPath("$HOME/.droid")
	viper.AutomaticEnv()

	if err := viper.BindEnv("rpc_endpoint", "RPC_ENDPOINT"); err != nil {
		return err
	}

	if err := viper.BindEnv("lcd_endpoint", "LCD_ENDPOINT"); err != nil {
		return err
	}

	if err := viper.ReadInConfig(); err != nil {
		return fmt.Errorf("Error reading config file: %s", err)
	}

	config.RPCEndpoint = viper.GetString("rpc_endpoint")
	config.LCDEndpoint = viper.GetString("lcd_endpoint")
	return nil
}

func main() {

	log.Info("🤖 droid is starting..")

	log.Info("Reading configuration.")
	if err := readConfig(); err != nil {
		log.Fatalf("Error reading configuration: %s", err)
	}
	log.Infof("RPC: %s", config.RPCEndpoint)
	log.Infof("LCD: %s", config.LCDEndpoint)
	log.Infof("Listening on port %s...", ":8080")

	statusURL = config.RPCEndpoint + "/status"
	epochURL = config.LCDEndpoint + "/osmosis/epochs/v1beta1/epochs"

	http.HandleFunc("/node_id", getNodeIDHandler)
	http.HandleFunc("/pub_key", getPubKeyHandler)
	http.HandleFunc("/block", getLatestBlockHandler)
	http.HandleFunc("/height", getLatestBlockHeightHandler)
	http.HandleFunc("/health", healthcheckHandler)

	if err := http.ListenAndServe(":8080", nil); err != nil {
		log.Fatalf("Fail to start server, %s\n", err)
	}
}


types.go
package main

import (
	"time"
)

func DefaultConfig() Config {
	return Config{
		RPCEndpoint: "http://0.0.0.0:26657",
		LCDEndpoint: "http://0.0.0.0:1317",
	}
}

type Config struct {
	RPCEndpoint string `mapstructure:"RPC_ENDPOINT"`
	LCDEndpoint string `mapstructure:"LCD_ENDPOINT"`
}

type NodeStatus struct {
	Result Result `json:"result"`
}

type Result struct {
	NodeInfo      NodeInfo      `json:"node_info"`
	SyncInfo      SyncInfo      `json:"sync_info"`
	ValidatorInfo ValidatorInfo `json:"validator_info"`
}

type SyncInfo struct {
	LatestAppHash       string    `json:"latest_app_hash"`
	LatestBlockHash     string    `json:"latest_block_hash"`
	LatestBlockHeight   string    `json:"latest_block_height"`
	LatestBlockTime     time.Time `json:"latest_block_time"`
	EarliestBlockHash   string    `json:"earliest_block_hash"`
	EarliestAppHash     string    `json:"earliest_app_hash"`
	EarliestBlockHeight string    `json:"earliest_block_height"`
	EarliestBlockTime   time.Time `json:"earliest_block_time"`
	CatchingUp          bool      `json:"catching_up"`
}

type NodeInfo struct {
	ID      string `json:"id"`
	Network string `json:"network"`
}

type ValidatorInfo struct {
	Address string `json:"address"`
	PubKey  PubKey `json:"pub_key"`
}

type PubKey struct {
	Type  string `json:"type"`
	Value string `json:"value"`
}

type Block struct {
	Hash   string    `json:"hash"`
	Height string    `json:"height"`
	Time   time.Time `json:"time"`
}


src/droid.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/spf13/viper"
)

var (
	config    Config
	statusURL = config.RPCEndpoint + "/status"
	epochURL  = config.LCDEndpoint + "/osmosis/epochs/v1beta1/epochs"
)

type Config struct {
	RPCEndpoint string `mapstructure:"RPC_ENDPOINT"`
	LCDEndpoint string `mapstructure:"LCD_ENDPOINT"`
}

type NodeStatus struct {
	Result Result `json:"result"`
}

type Result struct {
	NodeInfo      NodeInfo      `json:"node_info"`
	SyncInfo      SyncInfo      `json:"sync_info"`
	ValidatorInfo ValidatorInfo `json:"validator_info"`
}

type SyncInfo struct {
	LatestAppHash       string    `json:"latest_app_hash"`
	LatestBlockHash     string    `json:"latest_block_hash"`
	LatestBlockHeight   string    `json:"latest_block_height"`
	LatestBlockTime     time.Time `json:"latest_block_time"`
	EarliestBlockHash   string    `json:"earliest_block_hash"`
	EarliestAppHash     string    `json:"earliest_app_hash"`
	EarliestBlockHeight string    `json:"earliest_block_height"`
	EarliestBlockTime   time.Time `json:"earliest_block_time"`
	CatchingUp          bool      `json:"catching_up"`
}

type NodeInfo struct {
	ID      string `json:"id"`
	Network string `json:"network"`
}

type ValidatorInfo struct {
	Address string `json:"address"`
	PubKey  PubKey `json:"pub_key"`
}

type PubKey struct {
	Type  string `json:"type"`
	Value string `json:"value"`
}

type Block struct {
	Hash   string    `json:"hash"`
	Height string    `json:"height"`
	Time   time.Time `json:"time"`
}

func GetNodeStatus(statusURL string) NodeStatus {
	resp, err := http.Get(statusURL)
	if err != nil {
		log.Fatalln(err)
	}

	defer resp.Body.Close()
	var nodeStatus NodeStatus
	if err := json.NewDecoder(resp.Body).Decode(&nodeStatus); err != nil {
		log.Fatalln(err)
	}

	return nodeStatus
}

func getLatestBlockFromNodeStatus(status NodeStatus) Block {

	return Block{
		Height: status.Result.SyncInfo.LatestBlockHeight,
		Hash:   status.Result.SyncInfo.LatestBlockHash,
		Time:   status.Result.SyncInfo.LatestBlockTime,
	}
}

// /block handler
func getLatestBlockHandler(w http.ResponseWriter, r *http.Request) {

	status := GetNodeStatus(statusURL)
	latestBlock := getLatestBlockFromNodeStatus(status)

	w.Header().Set("Content-Type", "application/json")
	enc := json.NewEncoder(w)
	err := enc.Encode(latestBlock)

	if err != nil {
		log.Fatalf("Failed to encode block, %s\n", err)
		io.WriteString(w, "Internal Server Error\n")
	}
}

// /height handler
func getLatestBlockHeightHandler(w http.ResponseWriter, r *http.Request) {

	status := GetNodeStatus(statusURL)
	latestBlock := getLatestBlockFromNodeStatus(status)
	_, _ = io.WriteString(w, latestBlock.Height)
}

// /node_id handler
func getNodeIDHandler(w http.ResponseWriter, r *http.Request) {

	status := GetNodeStatus(statusURL)
	_, _ = io.WriteString(w, status.Result.NodeInfo.ID)
}

// /pub_key handler
func getPubKeyHandler(w http.ResponseWriter, r *http.Request) {
	status := GetNodeStatus(statusURL)

	response := map[string]string{
		"@type": "/cosmos.crypto.ed25519.PubKey",
		"key":   status.Result.ValidatorInfo.PubKey.Value,
	}

	data, _ := json.Marshal(response)
	_, _ = io.WriteString(w, string(data))
}

// global cached next epoch start time
var nextEpochStartTime time.Time

// This variable controls how many minutes after epoch we should expect the node to properly respond
// We assume epoch starts 5 minutes before the calculated epoch
var minutesAfterEpoch time.Duration = 35

func calculateEpoch() time.Time {
	failure := time.Time{}
	// Calculate the epoch from the node
	resp, err := http.Get(epochURL)
	if err != nil {
		log.Fatalln(err)
	}

	defer resp.Body.Close()
	bz, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println(err)
		return failure
	}
	if resp.StatusCode != http.StatusOK {
		fmt.Println(err)
		return failure
	}

	// Unmarshall the epoch json into a map from string to an array of interfaces
	var epoch map[string]interface{}
	err = json.Unmarshal(bz, &epoch)
	if err != nil {
		fmt.Println(err)
		return failure
	}

	// extract the "epoch" array from the epoch map
	epochArray := epoch["epochs"].([]interface{})

	// iterate over the array and find the element of the array that has identifier "day"
	for _, epochObject := range epochArray {
		epochObject := epochObject.(map[string]interface{})
		epochIdentifier := epochObject["identifier"].(string)
		//check that the identifier is "day"
		if epochIdentifier != "day" {
			continue
		}
		// extract the "current_epoch_start_time" from the epochObject and parse it as a date
		epochStartTimeString := epochObject["current_epoch_start_time"].(string)
		epochStartTime, err := time.Parse(time.RFC3339, epochStartTimeString)
		if err != nil {
			fmt.Println(err)
			return failure
		}
		// add 24 hours to the epoch start time
		nextEpochStartTime = epochStartTime.Add(24 * time.Hour)
		// subtract 5 minutes
		nextEpochStartTime = nextEpochStartTime.Add(-5 * time.Minute)
		return nextEpochStartTime
	}
	fmt.Println("No epochs found")
	return failure
}

// Function getEpoch2 which queries the epochURL and parses it as json
func isWithinEpoch(currentTime time.Time) bool {
	// Check if we have a cached epoch
	if nextEpochStartTime.IsZero() || currentTime.After(nextEpochStartTime.Add(minutesAfterEpoch*time.Minute)) {
		// if there is no cache or the epoch has ended, recalculate
		nextEpochStartTime = calculateEpoch()
	}
	// otherwise, use the cached value

	if currentTime.After(nextEpochStartTime) && currentTime.Before(nextEpochStartTime.Add(minutesAfterEpoch*time.Minute)) {
		return true
	}
	return false
}

// /health handler
func healthcheckHandler(w http.ResponseWriter, r *http.Request) {

	status := GetNodeStatus(statusURL)
	latestBlock := getLatestBlockFromNodeStatus(status)

	catchingUp := status.Result.SyncInfo.CatchingUp
	secondsSinceLastBlock := int(time.Now().UTC().Sub(latestBlock.Time).Seconds())

	// To test with fake epoch
	//timsStr := "2023-01-25T13:31:46.131312500Z"
	//t, _ := time.Parse(time.RFC3339Nano, timsStr)

	t := time.Now()
	isEpoch := isWithinEpoch(t)

	healthStatus := http.StatusOK
	healthMessage := "UP"

	if catchingUp || (secondsSinceLastBlock >= 60 && !isEpoch) {
		healthStatus = http.StatusServiceUnavailable
		healthMessage = "DOWN"
	}

	w.WriteHeader(healthStatus)
	_, _ = io.WriteString(w, fmt.Sprintf("%s\nLatest Block %s (Received %d seconds ago)\n", healthMessage, latestBlock.Height, secondsSinceLastBlock))
}

func setNodeEndpoints() error {
	viper.SetConfigName("config")
	viper.SetConfigType("yaml")
	viper.AddConfigPath("./")
	err := viper.ReadInConfig()
	if err != nil {
		return err
	}

	configUnmarshalled := Config{}
	err = viper.Unmarshal(&configUnmarshalled)
	if err != nil {
		return err
	}
	config = configUnmarshalled
	return nil
}

func main() {
	fmt.Println("[ 🤖 droid starting ]")

	// set config & endpoints
	fmt.Println("setting node endpoints...")
	err := setNodeEndpoints()
	if err != nil {
		panic("end points config are not set correctly")
	}
	fmt.Println("RPC end point: ", config.RPCEndpoint)
	fmt.Println("LCD end point: ", config.LCDEndpoint)

	fmt.Printf("listening on port %d...\n", 8080)

	http.HandleFunc("/node_id", getNodeIDHandler)
	http.HandleFunc("/pub_key", getPubKeyHandler)
	http.HandleFunc("/block", getLatestBlockHandler)
	http.HandleFunc("/height", getLatestBlockHeightHandler)
	http.HandleFunc("/health", healthcheckHandler)

	err = http.ListenAndServe(":8080", nil)
	if err != nil {
		log.Fatalf("Fail to start server, %s\n", err)
	}
}
